man/replcon.1
.TH replcon 1
.SH NAME
replcon \- SELinux file context replacement tool
.SH SYNOPSIS
.B replcon
NEW_CONTEXT DIR [OPTIONS] [EXPRESSION]
.SH DESCRIPTION
.PP
.B replcon
allows the user to find and replace file contexts.
Replacements can be filtered by object class as described below.
.SH REQUIRED ARGUMENTS
.IP NEW_CONTEXT
The replacement context as expressed as a partial context, described
below.
.IP DIR
Initial directory to begin searching.  The tool will recurse into any
subdirectories, so be sure there are no circular mounts within it.
.SH EXPRESSION
.P
The following options allow the user to specify which files to find.
A file must meet all specified criteria for its context to be
replaced.  If no expression is provided, all files' contexts are
replaced.
.IP "-t TYPE, --type=TYPE"
Search for files with a context containing the type TYPE.
.IP "-u USER, --user=USER"
Search for files with a context containing the user USER.
.IP "-r ROLE, --role=ROLE"
Search for files with a context containing the role ROLE.
.IP "-m RANGE, --mls-range=RANGE"
Search for files with a context with the MLS range of RANGE.  Note
that
.B replcon
ignores the SELinux translation library, if present.  In addition,
this flag is ignored if
.B DIR
has no MLS information.
.IP "--context=CONTEXT"
Search for files matching this partial context.  This flag overrides
-t, -u, -r, and -m.
.IP "-p PATH, --path=PATH"
Search for files which include PATH.
.IP "-c CLASS, --class=CLASS"
Search only files of object class CLASS.
.SH OPTIONS
.IP "-R, --regex"
Enable regular expressions
.IP "-v, --verbose"
Display context info during replacement.
.IP "-h, --help"
Print help information and exit.
.IP "-V, --version"
Print version information and exit.
.SH PARTIAL CONTEXT
The
.B --context
flag and
.B NEW_CONTEXT
argument specify a partial context, which is a a colon separated list
of user, role, and type.  If the system supports MLS, the context may
have a fourth field that gives the range.  With
.B --context
if a field is not specified or is the literal asterisk, then the query
will always match the field.  With
.B NEW_CONTEXT
if a field is not specified or is the literal asterisk, then that
portion of the context will not be modified.
.SH OBJECT CLASSES
Valid object class strings are
.PP
block,
char,
dir,
fifo,
file,
link, or
sock.
.SH NOTE
The replcon utility always operates on "raw" SELinux file contexts.
If the system has an installed translation library (i.e., libsetrans),
those translations are ignored in favor of reading the original
contexts from the filesystem.
.SH EXAMPLES
.TP
.B replcon ::type_t: .
Replace every context's type in the current directory with type_t.
The user and role portion remain unchanged.
.TP
.B replcon -u user_u *:role_r:* .
Replace every context's role with user user_u in the current directory
with role_r.  The user and type portion remain unchanged.
.TP
.B replcon --context ::type_t:s0 :::s0:c0 /tmp
Replace every context with type type_t and MLS range s0 in /tmp
with MLS range s0:c0.
.SH AUTHOR
This manual page was written by Jeremy A. Mowery <jmowery@tresys.com>.
.SH COPYRIGHT
Copyright(C) 2003-2007 Tresys Technology, LLC
.SH BUGS
Please report bugs via an email to setools-bugs@tresys.com.
.SH SEE ALSO
findcon(1), indexcon(1)


NEWS
SETools 3.3.6:

This release builds upon SETools 3.3.5:

  * Update attribute handling to use attributes in version 24 policy.

  * Fix bug where dontaudit rules were loaded when the "no rules" option is
    enabled.

seaudit: 

  * Add MLS fields to source and target contexts.

  * Double clicking a message line will display the orignal log message.

seinfo:

  * Qdd query for permissive types.

  * Add query for policy capabilities.

==================================================

SETools 3.3.5:

This release builds upon SETools 3.3.4:

  * Update to policy loader to match checkpolicy 2.0.16 and libsepol
    2.0.32.

  * Changes to libqpol to allow compiling against libsepol >= 2.0.29.

  * Support for reading policy version 23.

==================================================

SETools 3.3.4:

This release builds upon SETools 3.3.3:

  * Update to policy loader to match checkpolicy 2.0.13 and libsepol
    2.0.23.

  * Fixes to apol for proper handling of Tk 8.5.

  * Fixes to libapol, libqpol, and sechecker to build using GCC 4.3.

==================================================

SETools 3.3.3:

This release builds upon SETools 3.3.2:

  * Changes to libqpol to allow compiling against libsepol >= 2.0.20.

  * Support for reading policy version 22.

  * Clarification to default policy loading for seinfo, sesearch, and
    sechecker.

  * Build scripts for Debian and Ubuntu.

==================================================

SETools 3.3.2:

This release corrects a number of issues present in SETools 3.3.1:

  * Fix to libqpol for policies lacking genfscon statements; for
    policies containing disabled aliases; for
    qpol_type_get_alias_iter() given certain policies; for the special
    role object_r.

  * Fix to libapol when running a default avrule query on policies
    that have not had their neverallow rules loaded.

  * Fix to libsefs and apol for invalid regular expressions.

  * Fix to apol when validating empty levels.

  * Fix to all SWIG generated wrappers for Java.

==================================================

SETools 3.3.1:

This release corrects a number of issues present in SETools 3.3:

  * Fix to configure when Tcl is not found on build system; fix when
    compiling with --disable-gui option.

  * Fix to uninstall targets of Java wrappers.

  * Fix to libapol where transitive flows could return results that
    were supposed to be excluded; fixed memory leaks in infoflow graph
    generation.

  * Fix to libsefs when running MLS query on non-MLS fclists.

  * Fix to apol when reading older .apol files; fix copy and select
    all on certain tabs; fix to filter by attribute on some advanced
    dialogs.

==================================================

SETools 3.3:

SETools:

  * SETools now has an external dependency upon libsqlite3 >= 3.2.  The
    supplied configure script will enforce this dependency.

  * pkg-config scripts are installed with the SETools libraries.

libsefs:

  * Rewrite of library to have proper namespaces and much more usable
    object-oriented design.

  * SWIG wrappers generated for this library if the appropriate
    configure flags are set.

findcon, searchcon:

  * Merge searchcon's functionality into findcon.  The searchcon tool
    has been removed from SETools.

indexcon, replcon:

  * Updated to use new libsefs design.

apol:

  * Updated to use new libsefs design.

  * Modified to use the SWIG Tcl interface rather than a custom C
    library.  apol is now a combination of a Tcl script (simply called
    'apol') and associated packages that are required at runtime.

  * Neverallow rules are only loaded and expanded when the user
    performs a search for them.  This will dramatically speed up
    initial policy load time.

awish:

  * awish is no longer needed and thus has been removed from SETools.

sediff, sediffx:

  * Instead of differentiating "AV rules" or "TE rules", user now
    specifies which particular rule to compare (allow, dontaudit,
    type_transition, etc.).

  * Neverallow rules are only loaded and expanded when the user
    performs a diff upon them.  This will dramatically speed up
    initial policy load time.

==================================================

SETools 3.2:

libapol, libqpol, libesaudit, libpoldiff:

  * If --enable-swig-python is given during configure time, the build
    system will create Python SWIG wrappers for these libraries.

  * If --enable-swig-java is given during configure time, the build
    system will create Java SWIG wrappers for these libraries.

libpoldiff:

  * Provides ability to diff levels, categories, and range_transition
    statements.  Provides ability to diff modifications to a user's
    default level and permitted MLS range, assuming the policies are MLS.
    
apol, seaudit, sediffx:

  * Introduces 'policy list', a small text file that contains
    references to a base policy and any number of modules.  After
    selecting the base policy and modules in a tool's open policy
    dialog, click on 'Export' to write a policy list to disk.  That
    policy list then may be imported into the same tool or any other
    graphical SETools application.  This file also may be specified
    on the command line for all tools that load a policy.

sesearch, apol:

  * Provides full support for version 21 policy (i.e., object classes for
    range_transition statements).

sediff:

  * Shows MLS diffs.

sediffx:

  * Shows MLS diffs.

  * Provides option to show line numbers that contributed just to a
    specific AV rule's permission by clicking that AV rule's permissions.
    This is in addition to existing sediffx behavior that showed all lines
    associated with a particular rule difference.

  * Provides support for type joins and splits within the type remap dialog.

==================================================

SETools 3.1:

SETools:

  * All tools that open a policy now support loadable policy modules.

secmds/replcon.cc
/**
 * @file
 *
 * A tool for replacing file contexts in SELinux.
 *
 *  @author Jeremy A. Mowery jmowery@tresys.com
 *  @author Jason Tang jtang@tresys.com
 *
 * Copyright (C) 2003-2007 Tresys Technology, LLC
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <config.h>

#include <sefs/filesystem.hh>
#include <sefs/query.hh>
#include <selinux/selinux.h>
#include <apol/util.h>

using namespace std;

#include <assert.h>
#include <errno.h>
#include <getopt.h>
#include <iostream>
#include <stdlib.h>

#define COPYRIGHT_INFO "Copyright (C) 2003-2007 Tresys Technology, LLC"

enum OPTIONS
{
	OPTION_CONTEXT = 256
};

static struct option const longopts[] = {
	{"class", required_argument, NULL, 'c'},
	{"type", required_argument, NULL, 't'},
	{"user", required_argument, NULL, 'u'},
	{"role", required_argument, NULL, 'r'},
	{"mls-range", required_argument, NULL, 'm'},
	{"path", required_argument, NULL, 'p'},
	{"regex", no_argument, NULL, 'R'},
	{"context", required_argument, NULL, OPTION_CONTEXT},
	{"verbose", no_argument, NULL, 'v'},
	{"help", no_argument, NULL, 'h'},
	{"version", no_argument, NULL, 'V'},
	{NULL, 0, NULL, 0}
};

extern int lsetfilecon_raw(const char *, security_context_t) __attribute__ ((weak));

/**
 * As that setools must work with older libselinux versions that may
 * not have the _raw() functions, declare them as weak.  If libselinux
 * does indeed have the new functions then use them; otherwise
 * fallback to the originals.
 */
static int replcon_lsetfilecon(const char *path, security_context_t context)
{
	if (lsetfilecon_raw != NULL)
	{
		return lsetfilecon_raw(path, context);
	}
	else
	{
		return lsetfilecon(path, context);
	}
}

struct replcon_info
{
	bool verbose, mls;
	apol_context_t *replcon;
};

static void usage(const char *program_name, bool brief)
{
	cout << "Usage: " << program_name << " NEW_CONTEXT DIR [OPTIONS] [EXPRESSION]" << endl << endl;
	if (brief)
	{
		cout << "\tTry " << program_name << " --help for more help." << endl << endl;
		return;
	}

	cout << "Replace SELinux file contexts for files matching a given context." << endl << endl;

	cout << "REQUIRED ARGUMENTS :" << endl;
	cout << "  NEW_CONTEXT                    partial or full context to relabel" << endl;
	cout << "  DIR                            starting directory to replace" << endl;
	cout << endl;
	cout << "EXPRESSION:" << endl;
	cout << "  -t TYPE,  --type=TYPE          find contexts with type TYPE" << endl;
	cout << "  -u USER,  --user=USER          find contexts with user USER" << endl;
	cout << "  -r ROLE,  --role=ROLE          find contexts with role ROLE" << endl;
	cout << "  -m RANGE, --mls-range=RANGE    find contexts with MLS range RANGE" << endl;
	cout << "  --context=CONTEXT              partial or full context to find" << endl;
	cout << "                                 (overrides expression options above)" << endl;
	cout << "  -p PATH,  --path=PATH          find files in PATH" << endl;
	cout << "  -c CLASS, --class=CLASS        find files of object class CLASS" << endl;
	cout << endl;

	cout << "OPTIONS:" << endl;
	cout << "  -R, --regex                    enable regular expressions" << endl;
	cout << "  -v, --verbose                  show context of matching files" << endl;
	cout << "  -h, --help                     print this help text and exit" << endl;
	cout << "  -V, --version                  print version information and exit" << endl;
	cout << endl;
	cout << "If the fclist does not contain MLS ranges and -m was given," << endl;
	cout << "then the search will return nothing." << endl;
	cout << endl;
	cout << "NEW_CONTEXT is as a colon separated list of user, role, type, and MLS range" << endl;
	cout << "such as follows: user_u:object_r:user_t:s0.  If a field is not specified," << endl;
	cout << "that portion of the context will not be replaced." << endl;
	cout << "Examples:" << endl;
	cout << "    replcon ::type_t: ." << endl;
	cout << "        Replace all files and subdirectories in current directory with" << endl;
	cout << "        type type_t, recursing within the directory." << endl;
	cout << "    replcon -u user_u *:role_r:* ." << endl;
	cout << "        Replace files that contain user_u with role role_r." << endl;
	cout << "    replcon --context ::type_t:so :::s0:c0 /tmp" << endl;
	cout << "        Replace files with type type_t and level s0 in /tmp with MLS" << endl;
	cout << "        range s0:c0." << endl;
}

static int replace_entry(sefs_fclist * fclist, const sefs_entry * e, void *arg)
{
	struct replcon_info *r = static_cast < struct replcon_info *>(arg);
	const apol_context_t *scon = e->context();
	const char *user, *role, *type;
	char *con_str = NULL;
	size_t len = 0;

	// determine what the new context should be
	if ((user = apol_context_get_user(r->replcon)) == NULL)
	{
		user = apol_context_get_user(scon);
	}
	if ((role = apol_context_get_role(r->replcon)) == NULL)
	{
		role = apol_context_get_role(scon);
	}
	if ((type = apol_context_get_type(r->replcon)) == NULL)
	{
		type = apol_context_get_type(scon);
	}
	if (apol_str_appendf(&con_str, &len, "%s:%s:%s", user, role, type) < 0)
	{
		return -1;
	}
	if (r->mls)
	{
		const apol_mls_range_t *apol_range = NULL;
		char *range = NULL;
		if ((apol_range = apol_context_get_range(r->replcon)) == NULL)
		{
			apol_range = apol_context_get_range(scon);
		}
		if ((range = apol_mls_range_render(NULL, apol_range)) == NULL || apol_str_appendf(&con_str, &len, ":%s", range) < 0)
		{
			free(range);
			free(con_str);
			return -1;
		}
		free(range);
	}

	if (r->verbose)
	{
		char *lcon = NULL, *rcon = NULL;
		if (r->mls)
		{
			lcon = apol_context_render(NULL, r->replcon);
			rcon = apol_context_render(NULL, scon);
		}
		else
		{
			if (asprintf(&lcon, "%s:%s:%s",
				     apol_context_get_user(r->replcon),
				     apol_context_get_role(r->replcon), apol_context_get_type(r->replcon)) < 0)
			{
				lcon = NULL;
			}
			if (asprintf(&rcon, "%s:%s:%s",
				     apol_context_get_user(scon), apol_context_get_role(scon), apol_context_get_type(scon)) < 0)
			{
				rcon = NULL;
			}
		}
		if (lcon == NULL || rcon == NULL)
		{
			free(lcon);
			free(rcon);
			return -1;
		}
		printf("%s: %s --> %s\n", e->path(), lcon, rcon);
		free(lcon);
		free(rcon);
	}

	// until there is a way to create a security_context_t from a
	// char *, simply perform the implicit cast below
	if (replcon_lsetfilecon(e->path(), con_str) != 0)
	{
		cerr << "Could not set context " << con_str << " for file " << e->path() << "." << endl;
		free(con_str);
		return -1;
	}

	free(con_str);
	return 0;
}

int main(int argc, char *argv[])
{
	int optc;
	struct replcon_info r;

	r.verbose = false;
	r.replcon = NULL;
	sefs_query *query = new sefs_query();

	apol_context_t *context = NULL;
	try
	{
		while ((optc = getopt_long(argc, argv, "t:u:r:m:p:c:RvhV", longopts, NULL)) != -1)
		{
			switch (optc)
			{
			case 't':
				if (context == NULL)
				{
					query->type(optarg, false);
				}
				break;
			case 'u':
				if (context == NULL)
				{
					query->user(optarg);
				}
				break;
			case 'r':
				if (context == NULL)
				{
					query->role(optarg);
				}
				break;
			case 'm':
				if (context == NULL)
				{
					query->range(optarg, APOL_QUERY_EXACT);
				}
				break;
			case OPTION_CONTEXT:
				if ((context = apol_context_create_from_literal(optarg)) == NULL)
				{
					cerr << "Could not create source context." << endl;
					throw runtime_error(strerror(errno));
				}
				break;
			case 'p':
				query->path(optarg);
				break;
			case 'c':
				query->objectClass(optarg);
				break;
			case 'R':
				query->regex(true);
				break;
			case 'v':
				r.verbose = true;
				break;
			case 'h':     // help
				usage(argv[0], false);
				exit(0);
			case 'V':     // version
				cout << "replcon " << VERSION << endl << COPYRIGHT_INFO << endl;
				exit(0);
			default:
				usage(argv[0], true);
				exit(1);
			}
			if (context != NULL)
			{
				query->user(apol_context_get_user(context));
				query->role(apol_context_get_role(context));
				query->type(apol_context_get_type(context), false);
				if (apol_context_get_range(context) != NULL)
				{
					char *rng = apol_mls_range_render(NULL, apol_context_get_range(context));
					query->range(rng, APOL_QUERY_EXACT);
					free(rng);
				}
				else
				{
					query->range(NULL, APOL_QUERY_EXACT);
				}
			}
		}
	}
	catch(bad_alloc)
	{
		cerr << strerror(errno) << endl;
		apol_context_destroy(&context);
		delete query;
		exit(-1);
	}
	apol_context_destroy(&context);

	if (optind + 2 != argc)
	{
		usage(argv[0], 1);
		delete query;
		exit(-1);
	}

	sefs_fclist *fclist = NULL;
	try
	{
		fclist = new sefs_filesystem(argv[optind + 1], NULL, NULL);
		r.mls = fclist->isMLS();

		if ((r.replcon = apol_context_create_from_literal(argv[optind])) == NULL)
		{
			cerr << "Could not create replacement context." << endl;
			throw runtime_error(strerror(errno));
		}

		if (fclist->runQueryMap(query, replace_entry, &r) < 0)
		{
			throw runtime_error(strerror(errno));
		}
	}
	catch(...)
	{
		delete query;
		delete fclist;
		apol_context_destroy(&(r.replcon));
		exit(-1);
	}

	delete query;
	delete fclist;
	apol_context_destroy(&(r.replcon));
	return 0;
}
